addons/hr/models/resource.py

```diff
@@ -74,27 +69,65 @@ def _get_calendars_validity_within_period(self, start, end, default_company=None
-            )
-        resource_with_contract = self - resource_without_contract
-        if not resource_with_contract:
-            return calendars_within_period_per_resource
-        timezones = {resource.tz for resource in resource_with_contract}
+
+    def _get_contracts_valid_periods(self, start, end):
+        res = defaultdict(lambda: defaultdict(Intervals))
+        timezones = {resource.tz for resource in self}
         date_start = min(start.astimezone(timezone(tz)).date() for tz in timezones)
         date_end = max(end.astimezone(timezone(tz)).date() for tz in timezones)
-        contracts = resource_with_contract.employee_id._get_versions_with_contract_overlap_with_period(date_start, date_end)
+        contracts = self.employee_id._get_versions_with_contract_overlap_with_period(date_start, date_end)
         for contract in contracts:
             tz = timezone(contract.employee_id.tz)
-            calendars_within_period_per_resource[contract.employee_id.resource_id.id][contract.resource_calendar_id] |= Intervals([(
+            res[contract.employee_id.resource_id.id][contract.resource_calendar_id] |= Intervals([(
                 tz.localize(datetime.combine(contract.contract_date_start, datetime.min.time())) if contract.contract_date_start > start.astimezone(tz).date() else start,
                 tz.localize(datetime.combine(contract.contract_date_end, datetime.max.time())) if contract.contract_date_end and contract.contract_date_end < end.astimezone(tz).date() else end,
                 self.env['resource.calendar.attendance']
             )])
+        return res
+
+    def _get_calendars_validity_within_period(self, start, end, default_company=None):
+        assert start.tzinfo and end.tzinfo
+        if not self:
+            return super()._get_calendars_validity_within_period(start, end, default_company=default_company)
+        calendars_within_period_per_resource = defaultdict(lambda: defaultdict(Intervals))  # keys are [resource id:integer][calendar:self.env['resource.calendar']]
+        # Employees that have ever had an active contract
+        resource_without_contract = self._get_resource_without_contract()
+        if resource_without_contract:
+            calendars_within_period_per_resource.update(
+                super(ResourceResource, resource_without_contract)._get_calendars_validity_within_period(start, end, default_company=default_company)
+            )
+        resource_with_contract = self - resource_without_contract
+        if not resource_with_contract:
+            return calendars_within_period_per_resource
+
+        calendars_within_period_per_resource.update(resource_with_contract._get_contracts_valid_periods(start, end))
+        return calendars_within_period_per_resource
+
+    ####################### Section Methods only for planning and project apps #######################
+
```

Comment:
you can remove those comments, we will directly see your added methods are used in services app by grepping the methods name. 🙂